vortex-array/src/elementwise.rs

```diff
@@ -0,0 +1,123 @@
+use vortex_dtype::{DType, NativePType, PType};
+use vortex_error::{VortexError, VortexResult};
+use vortex_scalar::Scalar;
+
+use crate::iter::Batch;
+use crate::{Array, ArrayDType};
+
+pub enum OtherValue {
+    Scalar(Scalar),
+    Array(Array),
+}
+
+impl OtherValue {
+    pub fn dtype(&self) -> &DType {
+        match self {
+            Self::Scalar(s) => s.dtype(),
+            Self::Array(a) => a.dtype(),
+        }
+    }
+}
+
+impl From<Array> for OtherValue {
+    fn from(value: Array) -> Self {
+        Self::Array(value)
+    }
+}
+
+impl From<Scalar> for OtherValue {
+    fn from(value: Scalar) -> Self {
+        Self::Scalar(value)
+    }
+}
+
+pub trait BinaryFn {
+    fn binary<
+        I: NativePType + TryFrom<Scalar, Error = VortexError>,
```

Comment:
this TryFrom is unnecessary fwiw, same for UnaryFn

Reply:
as always, belonged to a previous iteration